[fips] Propagate the FIPS boolean to the manager, raft storage layer, and raft DEK management

cyli opened the pull request:
This PR ~~is based on https://github.com/docker/swarmkit/pull/2562, and~~ propagates the FIPS boolean to other parts of the manager that need it.  This also documents the `KeyReadWriter` and `RaftDEKManager` better, so that reviewers can understand how they interact.

~~To view these changes alone, please see https://github.com/docker/swarmkit/pull/2586/files/b2da71c62612e2109d563fa05a764284caf0ec74..7003b358d8051bce2615d2ae5cd1e30c585250b4.~~ (no need, #2562 was merged)

The code changes are actually fairly small ~~- most of the change is comments and a [prose doc](https://github.com/cyli/swarmkit/blob/b2ec3d83c3523fe593121bf76a46482ca7d75170/design/raft_encryption.md) about raft encryption~~.  There are also some test changes to make sure that the fips boolean is handled properly, and a refactor to make one particular test more readable (table-driven).

Docs for how raft encryption works, which may help reviewers, has been cherry-picked into https://github.com/docker/swarmkit/pull/2622

cyli commented:
Rebased!  The [first commit](https://github.com/docker/swarmkit/pull/2586/commits/d6ad6db936f4d1bda5e0c11b166e0319d910fdff) is the main change, and a lot of it is a test refactor.

The other two changes are:

1.  a lot more documentation for raft encryption, so that reviewers can understand the first commit
2. propagating the FIPS var from the node to the manager

This should be ready for review now.

anshulpundir commented on manager/deks.go:
```diff
@@ -112,6 +118,7 @@ func compareKEKs(oldKEK, candidateKEK ca.KEKData) (bool, bool, error) {
 type RaftDEKManager struct {
 	kw         ca.KeyWriter
 	rotationCh chan struct{}
+	FIPS       bool
```
nit: add a comment on what this flag indicates.

anshulpundir commented on manager/deks.go:
```diff
@@ -156,8 +165,9 @@ func (r *RaftDEKManager) NeedsRotation() bool {
 // GetKeys returns the current set of DEKs.  If NeedsRotation is true, and there
-// is no existing PendingDEK, it will try to create one.  If there are any errors
-// doing so, just return the original.
+// is no existing PendingDEK, it will try to create one.  If it successfully creates
```
The name 'GetKeys' implies a read-only operation so I would suggest avoid doing operations within this function that have side effects.

cyli replied:
I'd be happy to revisit this at a later point, but that's probably beyond the scope of this PR.

anshulpundir commented on manager/manager.go:
```diff
@@ -121,6 +121,9 @@ type Config struct {
 	// PluginGetter provides access to docker's plugin inventory.
 	PluginGetter plugingetter.PluginGetter
+
+	// FIPS specifies whether the manager should run in FIPS-compliant mode.
```
nit: does it make sense to say that the mode is specified during startup and can't be changed ?

dperny commented:
To confirm, the diff on the test is only huge because it basically takes every test and wraps them in a loop to check both fips and non-fips, right? 

LGTM.

cyli commented:
@dperny Yes pretty much.  For `TestRaftDEKManagerNeedsRotateGetKeys` I also refactored it to be a table-driven test, but wrapped it in a loop to test both FIPS and non-FIPS as well.

cyli commented:
Thanks @anshulpundir.  I've think I've addressed your comments, except for https://github.com/docker/swarmkit/pull/2586/files/04230609c0ddb58a391bfa9defb837f6ac532299#diff-c837f367ebb298ae3e59a7a95ab64429, since that'd be a pretty big change and we might want to do that in another PR.  I'd also have to think about it some more - it's mutating to avoid having to do a bunch of extra rotations.

anshulpundir commented:
LGTM

dperny commented:
Needs a rebase, then LGTM
